pkg/daemon/controller/controller.go

```diff
@@ -265,6 +276,86 @@ func (sc *ServiceController) Reconcile(_ context.Context, req ctrl.Request) (res
 }
 
+// reconcileEnvironmentVariables makes sure that the proxy variables exist as expected, or are safely rectified.
+// Returns a boolean expressing whether the instance is awaiting a reboot.
+func (sc *ServiceController) reconcileEnvironmentVariables(envVars map[string]string, node core.Node) (bool, error) {
+	restartRequired, err := sc.ensureVarsAreUpToDate(envVars)
+	if err != nil {
+		return false, err
+	}
+	if restartRequired {
+		// Applying the reboot annotation results in an event picked up by WMCO's node controller to reboot the instance
+		if err = metadata.ApplyRebootAnnotation(sc.ctx, sc.client, node); err != nil {
+			return false, fmt.Errorf("error setting reboot annotation on node %s: %w", sc.nodeName, err)
+		}
+		return true, nil
+	}
+	// Wait until the environment variables at the process level are set as expected.
+	// This will only be picked up after WMCO reboots the instance
+	err = wait.PollUntilContextTimeout(sc.ctx, 15*time.Second, 5*time.Minute, true,
+		func(ctx context.Context) (done bool, err error) {
+			stillNeedsReboot := false
+			for _, varName := range cluster.SupportedProxyVars {
+				cmd := fmt.Sprintf("[Environment]::GetEnvironmentVariable('%s', 'Process')", varName)
+				out, err := sc.psCmdRunner.Run(cmd)
+				if err != nil {
+					return false, fmt.Errorf("error running PowerShell command %s with output %s: %w", cmd, out, err)
+				}
+				if strings.TrimSpace(out) != envVars[varName] {
+					stillNeedsReboot = true
+				}
+			}
+			return !stillNeedsReboot, nil
+		})
+	if err != nil {
+		return true, fmt.Errorf("error waiting for environment vars to get picked up by processes: %w", err)
+	}
+	// Remove the reboot annotation after we know the reboot occurred successfully. No-op if already not present
+	return false, metadata.RemoveRebootAnnotation(sc.ctx, sc.client, node)
+}
+
+// ensureVarsAreUpToDate ensures that the proxy environment variables are set as expected on the instance
+// If there's any changes, an instance restart is required to ensure all processes pick up the updated values.
+func (sc *ServiceController) ensureVarsAreUpToDate(envVars map[string]string) (restartRequired bool, err error) {
+	// systemEnvVarRegistryPath is where system level environment variables are stored in the Windows OS
+	const systemEnvVarRegistryPath = `SYSTEM\CurrentControlSet\Control\Session Manager\Environment`
+	for key, expectedVal := range envVars {
+		registryKey, err := registry.OpenKey(registry.LOCAL_MACHINE, systemEnvVarRegistryPath, registry.ALL_ACCESS)
+		if err != nil {
+			return false, fmt.Errorf("unable to open Windows system registry key %s: %w", systemEnvVarRegistryPath, err)
+		}
+		defer func() { // always close the registry key, without swallowing any error returned before the defer call
+			closeErr := registryKey.Close()
+			if closeErr != nil {
+				if err != nil {
+					err = fmt.Errorf("multiple errors: %w, %w", err, closeErr)
+				} else {
+					err = closeErr
+				}
+			}
+		}()
```

Comment:
Why does this need to be done this way and not just logged?
In case of this error, the handle is being leaked no matter what, triggering another reconcile is not going to fix that.

Reply:
Taking the straightforward approach and just logging it shouldn't contradict @aravindhp's comment. The error is still being noticed.
https://github.com/openshift/windows-machine-config-operator/pull/1536#discussion_r1258583602

Reply:
> another reconcile is not going to fix that

Do we know this for sure? It could be that next time we reconcile the handle can successfully closes